src/data/syncset/path.rs

```diff
@@ -0,0 +1,429 @@
+use std::convert::TryInto;
+
+use crate::crypto::HashError;
+use crate::crypto::hash::{hash, Digest, SIZE as HASH_SIZE};
+use super::errors::PathLengthError;
+
+use serde::{Serialize, Deserialize};
+
+const BITS_IN_BYTE: usize = 8;
+
+/// Navigator wrapper for Digest
+/// Guaranteed to have HASH_SIZE * 8 bits of depth
+#[derive(Clone, Debug, PartialEq)]
+pub struct Path(pub(super) Digest);
+
+/// Navigator
+/// Guaranteed to have 0 <= n <= HASH_SIZE * 8 bits of depth
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Prefix {
+    inner: [u8; HASH_SIZE],
+    depth: usize,
+}
+
+/// Direction enumeration for abstraction of bit navigation.
+/// 0 is Left, 1 is Right
+#[derive(Eq, PartialEq, Debug)]
+pub enum Direction {
+    Left,
+    Right,
+}
+
+impl Direction {
+    /// Convert the i-th bit of the byte into a Direction
+    pub fn from_bit(byte: u8, bit_idx: usize) -> Direction {
+        if is_bit_set(byte, bit_idx) {
+            Direction::Right
+        } else {
+            Direction::Left
+        }
+    }
+
+    /// Convert the Direction to a bit
+    pub fn to_bit(&self) -> bool {
+        self == &Direction::Right
+    }
+}
+
+impl Path {
+    /// The number of bits in a hash digest
+    pub const NUM_BITS: usize = HASH_SIZE * BITS_IN_BYTE;
+
+    /// Returns the direction at a given bit index
+    /// Note that this function will panic if given an index
+    /// greater or equal to the number of bits in a hash digest
+    pub fn at(&self, idx: usize) -> Result<Direction, PathLengthError> {
+        if idx < Path::NUM_BITS {
+            let (byte_idx, bit_idx) = split_bits(idx);
+
+            debug_assert!(HASH_SIZE > byte_idx as usize, "Out of bounds byte index");
```

Comment:
I'd rather avoid having asserts outside of tests and return Result instead in public functions